File: backend/Models/Quiz.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace prid_2324_a02.Models;

public class Quiz
{
	[Key]
	public int Id { get; set; }
	public string Name { get; set; } = null!;
	public string? Description { get; set; }
	public int DatabaseId { get; set; }
	public Database Database { get; set; } = null!;
	public bool IsPublished { get; set; }
	public bool IsClosed { get; set; } = false;
	public bool IsTest { get; set; } = false;
	public DateTimeOffset? Start { get; set; }
	public DateTimeOffset? Finish { get; set; }

	public virtual ICollection<Attempt> Attempts { get; set; } = new HashSet<Attempt>();
	public virtual ICollection<Question> Questions { get; set; } = new HashSet<Question>();

	[NotMapped]
	public string? Status { get; set; } = null;
	[NotMapped]
	public bool Editable { get; set; } = true;
	[NotMapped]
	public int FirstQuestionId { get; set; }
	[NotMapped]
	public string? Evaluation { get; set; }

	public Quiz AddStatus(Attempt? attempt) {

		if (IsTest) {
			var today = DateTime.Now;
			if (Start!.Value > today) {
				Status = "A VENIR";
				return this;
			}
			if (Finish!.Value < today)
				Status = "CLOTURE";
			else
				SetStatusForAttempt(attempt);
			return this;
		}
		SetStatusForAttempt(attempt);
		return this;
	}

	private void SetStatusForAttempt(Attempt? attempt) {
		if (attempt == null) {
			Status = "PAS COMMENCE";
			return ;
		}
		if (attempt.Finish != null)
			Status = "FINI";
		else
			Status = "EN COURS";
	}
}

File: frontend/src/app/models/quiz.ts
import { Type } from "class-transformer";
import 'reflect-metadata';
import { Database } from "./database";
import { Question } from "./question";

export class Quiz {
	id?: number;
	name?: string;
	description?: string;
	isPublished?: boolean;
	isClosed?: boolean;
	isTest?: boolean;
	@Type(() => Date)
	start?: Date;
	@Type(() => Date)
	finish?: Date;
	database?: Database;
	status?: string;
	editable?: boolean;
	firstQuestionId: number = 0;
	evaluation?: string;

	questions: Question[] = [];

	get display(): string {
		return `${this.name} - (${this.description ? this.description : 'no description'})`;
	}

	get evaluationAsString(): string {
		return this.evaluation ?? "N/A";
	}

	get getTestOrTraining(): string {
		return this.isTest ? "Test" : "Training";
	}

	get dbName(): string | undefined {
		return this.database?.name;
	}

	get startAsString(): string {
		return this.start ? this.start.toLocaleDateString('eu-EU') : "N/A";
	}

	get finishAsString(): string {
		return this.finish ? this.finish.toLocaleDateString('eu-EU') : "N/A";
	}
}
